### PR TITLE
feat: add support to interpolate and fill in custom template with values

### PR DESCRIPTION
**Pull request by moaazsidat**
**Proposed changes**:
- Addresses part of https://github.com/RasaHQ/rasa/issues/3763
- Adds support to interpolate and fill in `custom` templates with slot values and passed values

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

**Comment by msamogh** [comment]
Thanks for the PR, we'll give it a review as soon as possible.

File: rasa/core/nlg/template.py [hunk 1]
```diff
@@ -104,25 +104,33 @@ def _fill_template_text(
...
-                    "without filling the template. "
-                    "".format(template, e.args[0])
-                )
+        if template_vars:
+            if "text" in template:
+                try:
+                    template["text"] = interpolate_text(template["text"], template_vars)
+                except KeyError as e:
+                    logger.exception(
+                        "Failed to fill utterance template '{}'. "
```
Comment (tabergma) [inline rasa/core/nlg/template.py]: As we are throwing the exact same exception for both template types, I would suggest to either move the try/except up one level or put the exception message into a constant.

File: tests/core/test_nlg.py [hunk 2]
```diff
@@ -109,17 +109,58 @@ def test_nlg_schema_validation_empty_image():
...
+        ("tag.with.dot", "chocolate"),
+        ("tag-w-dash", "apple pie"),
+        ("tag-w-$", "banana"),
+        ("tag-w-@", "one"),
+        ("tagCamelCase", "two"),
+        ("tag-w-*", "three"),
+        ("tag_w_underscore", "a"),
+        ("tag.with.float.val", 1.3),
+        ("tag-w-$", "banana"),
+        ("tagCamelCase", "two"),
```
Comment (tabergma) [inline tests/core/test_nlg.py]: Can you maybe add some edge cases like `None` or empty string?

File: rasa/core/nlg/interpolator.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import re
+
+
+def interpolate_text(text, values):
```
Comment (tabergma) [inline rasa/core/nlg/interpolator.py]: I'm not so happy with the name. Maybe something like `transforming_text_template` and `transforming_custom_template`?

Comment (moaazsidat) [reply rasa/core/nlg/interpolator.py]: sure thing, happy to update, got the name from 'string interpolation' (https://en.wikipedia.org/wiki/String_interpolation)

Comment (tabergma) [reply rasa/core/nlg/interpolator.py]: Did not know about that. Then we should maybe keep it. Sorry.

File: tests/core/test_nlg.py (hunk 2 above)
Comment (moaazsidat) [reply tests/core/test_nlg.py]: @tabergma thoughts on how `None` should be treated? If we stringify it, it'll just be turned into `"None"` which isn't what we'd want?

File: rasa/core/nlg/interpolator.py (hunk 3 above)
Comment (moaazsidat) [reply rasa/core/nlg/interpolator.py]: All good! Thanks for the feedback!

File: tests/core/test_nlg.py (hunk 2 above)
Comment (moaazsidat) [reply tests/core/test_nlg.py]: added a few edge cases to the tests
